### PR TITLE
Add allow_reassign plugin

### LINKED ISSUE #420: Error with non cached JSON/Hash value in v3.0.0
Hi, thanks for this gem, it is amazing!

After updating to version 3.0.0, I'm getting an error when trying to update an existing record using JSON/Hash as value.

In versions before 3.0, the assign method checks for cached files:

```ruby
def assign(value, **options)
  if value.is_a?(String)
    return if value == "" || !cached?(uploaded_file(value))
    assign_cached(uploaded_file(value))
  else
    uploaded_file = cache!(value, action: :cache, **options) if value
    set(uploaded_file)
  end
end
```

While in version 3.0, the assign method treats String/Hash values always as cached files:

```ruby
def attach_cached(value, **options)
  if value.is_a?(String) || value.is_a?(Hash)
    change(cached(value, **options), **options)
  else
    attach(value, storage: cache_key, action: :cache, **options)
  end
end
```

```
/home/jonian/.gem/ruby/2.6.0/gems/shrine-3.0.0/lib/shrine/attacher.rb:355:in `cached': expected cached file, got "{\"id\":\"534e009d3701166238fe95557b737661.jpg\",\"storage\":\"store\",\"metadata\":{\"size\":44399,\"color\":\"#091C26\",\"width\":1920,\"height\":1080,\"palette\":[\"#091C26\",\"#2C4B56\",\"#1E3A46\",\"#4D6A74\"],\"filename\":\"Rain.jpg\",\"mime_type\":\"image/jpeg\"}}" (Shrine::Error)
```

### PR DESCRIPTION
**Pull request by janko**
When coupled with a SPA, some users want to treat the attachment attribute as a permanent resource attribute, meaning that the API will always return attached file data (whether it's cached or stored) and the frontend will send the attached file data back on resource updates.

This was possible with Shrine 2.x, as then the attacher would simply skip any non-cached files. This changed in Shrine 3.0, where attempting to assign a non-cached file will now raise an exception, to let you know the operation didn't succeed.

However, for some users this feature was convenient as it meant less boilerplate when building a SPA + API. So, we add it back in form of a plugin. It makes the assignment like the following a no-op:

```rb
Shrine.plugin :allow_reassign
```
```rb
photo.image #=> #<Shrine::UploadedFile id="foo" storage=:store ...>
photo.image = '{"id":"foo","storage":"store",...}' # no-op
```

The plugin implementation is not as efficient as it could be; it will parse the cached attached file JSON data twice on assignment. However, avoiding the double-parse would require some non-trivial changes to the base API which I don't think are worth it at the moment.

Fixes #420

cc @Sexual @jonian @shrinerb/core

### CODE REVIEW

File: test/plugin/allow_assign_current_test.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  before do
+    @attacher = attacher { plugin :allow_assign_current }
+    @shrine   = @attacher.shrine_class
+  end
+
+  it "ignores assigning current non-cached file" do
+    file = @attacher.upload(fakeio, :other_store)
+    @attacher.file = file
+    @attacher.assign(file.to_json)
+    assert_equal file , @attacher.file
```
Comment (hmistry) [inline test/plugin/allow_assign_current_test.rb]: This test shows the file is the same but not sure if it tests whether it was due to it being skipped or uploading the same file. Same for test on L24 below.

**Comment by Sexual** [comment]
+1 on the name not really being self-explanatory. 

Maybe something like the following, but they're still not great:
allow_reassign
ignore_reassign_exception

File: test/plugin/allow_assign_current_test.rb (hunk 1 above)
Comment (janko) [reply test/plugin/allow_assign_current_test.rb]: This test asserts that no exception is raised, while the test below asserts that metadata is not changing. These two tests together should assert we're "skipping" the assignment (along with the `restore_cached_data` test at the bottom). If there is some other behavior you think should be asserted, let me know.

**Comment by janko** [comment]
@Sexual I like `allow_reassign`, it's shorter and I also think it communicates better than `allow_assign_current` 👍

**Comment by janko** [comment]
My only issue with using "reassign" is that the definition of that word to me implies *changing* something, and changing attachments was already allowed; it's the no-op assignment (or assigning "back") that we're supporting.

I know it would be easier if this was supported by default, as then we wouldn't have to come up with a name 😄, but the cognitive overhead when writing this feature was enough to convince me that it should remain a plugin.

I'm still thinking of something better 🤔

**Comment by aglushkov** [comment]
It would be great if this plugin allows to update custom metadata attributes, keeping same file. I would be using it to update cropping coordinates of already uploaded image. I have custom plugin for it, that checks file is same in similar manner

**Comment by jrochkind** [comment]
@aglushkov I believe you can do that in Shrine 3.0 without any plugin at all -- although I'm forgetting/not finding the exact API to do it.  (I too ran into it being hard in shrine 2.x). Maybe ask on the [discourse forum](https://discourse.shrinerb.com/).

**Comment by hmistry** [comment]
@janko I agree with reassign. I like `allow_persistent_attribute` better because it's more about the attribute though it's a bit too general like we don't allow persistence without it. See if we can capture in particular the skip-persistence-if-same part unless that's too specific.

**Pull request merged**